src/main/java/se/sundsvall/billingdatacollector/integration/opene/util/XPathUtil.java

```diff
@@ -22,127 +22,127 @@
-            .filter(not(Elements::isEmpty))
-            .map(Elements::getFirst)
-            .map(Element::text);
-    }
-
-    public static Document parseXmlDocument(final byte[] xml) {
-        return parse(new String(xml, StandardCharsets.ISO_8859_1), xmlParser());
-    }
-
-    public static Elements evaluateXPath(final byte[] xml, final String expression) {
-        var doc = parseXmlDocument(xml);
-
-        return Xsoup.compile(expression).evaluate(doc).getElements();
-    }
-
-    public static Elements evaluateXPath(final Element element, final String expression) {
-        return Xsoup.compile(expression).evaluate(element).getElements();
-    }
-
-    record Parameter(Field field, Class<?> type, Object value) { }
+	private XPathUtil() {}
+
+	public static <T> T extractValue(final byte[] xml, final Class<T> targetClass) {
+		// Check if objects of the given target class can be instantiated
+		if (isAbstract(targetClass.getModifiers()) || isInterface(targetClass.getModifiers())) {
+			throw new XPathException("%s must be a concrete class or a record".formatted(targetClass.getName()));
+		}
+
+		// Get the fields on the target class
+		var fields = targetClass.getDeclaredFields();
+		// Use arrays to ensure proper parameter ordering for records
+		var parameters = new Parameter[fields.length];
+		// Process the fields
+		for (var i = 0; i < fields.length; i++) {
+			var field = fields[i];
+			// Get the path annotation for the field
+			var pathAnnotation = AnnotationUtils.getAnnotation(field, XPath.class);
+			// We can't do anything with the field if the annotation is missing - bail out
+			if (pathAnnotation == null) {
+				continue;
+			}
+
+			var type = field.getType();
+			var path = pathAnnotation.value();
+
+			var value = getValue(xml, path, type);
+
+			parameters[i] = new Parameter(field, type, value);
+		}
+
+		try {
+			if (targetClass.isRecord()) {
+				var parameterTypes = new Class[parameters.length];
+				var parameterValues = new Object[parameters.length];
+				for (var i = 0; i < parameters.length; i++) {
+					parameterTypes[i] = parameters[i].type;
+					parameterValues[i] = parameters[i].value;
+				}
+
+				var constructor = targetClass.getDeclaredConstructor(parameterTypes);
+				constructor.setAccessible(true);
+				return constructor.newInstance(parameterValues);
+			} else {
+				var constructor = targetClass.getDeclaredConstructor();
+				constructor.setAccessible(true);
+				var result = constructor.newInstance();
+
+				for (var parameter : parameters) {
+					if (parameter != null) {
+						parameter.field.setAccessible(true);
+						parameter.field.set(result, parameter.value);
+					}
+				}
+
+				return result;
+			}
+		} catch (Exception e) {
+			throw new XPathException("Unable to extract value", e);
+		}
+	}
+
+	public static <T> T getValue(final byte[] xml, final String path, final Class<T> type) {
+		Object value;
+
+		if (type.equals(String.class)) {
```

Comment:
switch case?

Reply:
If `type` was an object it would have worked fine, now unfortunately not :)